bfv/bfv_test.go

```diff
@@ -227,413 +228,607 @@ func testEncoder(testctx *testContext, t *testing.T) {
-		values2, _, ciphertext2 := newTestVectorsRingQ(testctx, testctx.encryptorPk, t)
+		tc.evaluator.QuantizeToLvl(1, ciphertext1)
+		tc.evaluator.QuantizeToLvl(1, ciphertext2)
 
-		testctx.evaluator.AddNoMod(ciphertext1, ciphertext2, ciphertext1)
-		testctx.evaluator.Reduce(ciphertext1, ciphertext1)
-		testctx.ringT.Add(values1, values2, values1)
+		tc.evaluator.Add(ciphertext1, ciphertext2, ciphertext1)
+		tc.ringT.Add(values1, values2, values1)
 
-		verifyTestVectors(testctx, testctx.decryptor, values1, ciphertext1, t)
+		verifyTestVectors(tc, tc.decryptor, values1, ciphertext1, t)
 	})
 
-	t.Run(testString("Evaluator/AddNew/op1=Ciphertext/op2=Ciphertext", testctx.params), func(t *testing.T) {
+	t.Run(testString("QuantizeToLvl/MulRelin", tc.params), func(t *testing.T) {
 
-		values1, _, ciphertext1 := newTestVectorsRingQ(testctx, testctx.encryptorPk, t)
-		values2, _, ciphertext2 := newTestVectorsRingQ(testctx, testctx.encryptorPk, t)
+		values1, _, ciphertext1 := newTestVectorsRingQ(tc, tc.encryptorPk, t)
+		values2, _, ciphertext2 := newTestVectorsRingQ(tc, tc.encryptorPk, t)
 
-		ciphertext1 = testctx.evaluator.AddNew(ciphertext1, ciphertext2)
-		testctx.ringT.Add(values1, values2, values1)
+		tc.evaluator.QuantizeToLvl(1, ciphertext1)
+		tc.evaluator.QuantizeToLvl(1, ciphertext2)
 
-		verifyTestVectors(testctx, testctx.decryptor, values1, ciphertext1, t)
+		receiver := NewCiphertext(tc.params, ciphertext1.Degree()+ciphertext2.Degree())
+		tc.evaluator.Mul(ciphertext1, ciphertext2, receiver)
+		tc.ringT.MulCoeffs(values1, values2, values1)
+
+		verifyTestVectors(tc, tc.decryptor, values1, receiver, t)
+
+		receiver2 := tc.evaluator.RelinearizeNew(receiver)
+		verifyTestVectors(tc, tc.decryptor, values1, receiver2, t)
 	})
 
-	t.Run(testString("Evaluator/AddNoModNew/op1=Ciphertext/op2=Ciphertext", testctx.params), func(t *testing.T) {
+	t.Run(testString("QuantizeToLvl/Rotate", tc.params), func(t *testing.T) {
+
+		values1, _, ciphertext1 := newTestVectorsRingQ(tc, tc.encryptorPk, t)
 
-		values1, _, ciphertext1 := newTestVectorsRingQ(testctx, testctx.encryptorPk, t)
-		values2, _, ciphertext2 := newTestVectorsRingQ(testctx, testctx.encryptorPk, t)
+		tc.evaluator.QuantizeToLvl(1, ciphertext1)
 
-		ciphertext1 = testctx.evaluator.AddNoModNew(ciphertext1, ciphertext2)
-		ciphertext1 = testctx.evaluator.ReduceNew(ciphertext1)
-		testctx.ringT.Add(values1, values2, values1)
+		rotkey := tc.kgen.GenRotationKeysForRotations(nil, true, tc.sk)
+		evaluator := tc.evaluator.WithKey(rlwe.EvaluationKey{Rlk: tc.rlk, Rtks: rotkey})
 
-		verifyTestVectors(testctx, testctx.decryptor, values1, ciphertext1, t)
+		tc.evaluator.QuantizeToLvl(1, ciphertext1)
+
+		evaluator.RotateRows(ciphertext1, ciphertext1)
+		values1.Coeffs[0] = append(values1.Coeffs[0][tc.params.N()>>1:], values1.Coeffs[0][:tc.params.N()>>1]...)
+		verifyTestVectors(tc, tc.decryptor, values1, ciphertext1, t)
 	})
+}
+
+func testEvaluator(tc *testContext, t *testing.T) {
```

Comment:
One problem I see with the current test suite is that we do not test all operations at lower level. This same fact has caused several bugs and issues for the `ckks` package so we shouldn't repeat the same mistake again. To be discussed.

Reply:
Done